src/Enzo/enzo_EnzoEOSIdeal.cpp

```diff
@@ -0,0 +1,409 @@
+  etot = array_factory.from_grouping(integrable_group, "total_energy", 0);
+  if (idual){
+    eint = array_factory.from_grouping(integrable_group, "internal_energy", 0);
+  }
+  bx = array_factory.from_grouping(integrable_group, "bfield", 0);
+  by = array_factory.from_grouping(integrable_group, "bfield", 1);
+  bz = array_factory.from_grouping(integrable_group, "bfield", 2);
+
+  float ggm1 = get_gamma()*(get_gamma() - 1.);
+  enzo_float pressure_floor = get_pressure_floor();
+  enzo_float inv_gm1 = 1./(get_gamma()-1.);
+
+  // a requirement for an element of the internal energy field, cur_eint,
+  // to be updated to the value computed from the total energy field, eint_1,
+  // is that cur_eint > half_factor * cur_eint, where half_factor is 0.5. To
+  // allow eta = 0, to specify that this update should always occur, we set
+  // half_factor = 0 when eta = 0. 
+  const double half_factor = (eta != 0.) ? 0.5 : 0.;
+
+  for (int iz=0; iz<density.shape(0); iz++) {
+    for (int iy=0; iy<density.shape(1); iy++) {
+      for (int ix=0; ix<density.shape(2); ix++) {
+
+	enzo_float inv_rho = 1./density(iz,iy,ix);
+	enzo_float eint_floor = pressure_floor*inv_gm1*inv_rho;
+
+	enzo_float v2 = (vx(iz,iy,ix) * vx(iz,iy,ix) +
+			 vy(iz,iy,ix) * vy(iz,iy,ix) +
+			 vz(iz,iy,ix) * vz(iz,iy,ix));
+	enzo_float b2 = (bx(iz,iy,ix) * bx(iz,iy,ix) +
+			 by(iz,iy,ix) * by(iz,iy,ix) +
+			 bz(iz,iy,ix) * bz(iz,iy,ix));
+	enzo_float kinetic = 0.5*v2;
+	enzo_float magnetic = 0.5 * b2 *inv_rho;
+
+	if (idual){
+	  enzo_float eint_1 = etot(iz,iy,ix) - kinetic - magnetic;
+	  enzo_float cur_eint = eint(iz,iy,ix);
+
+	  // compute cs^2 with estimate of eint from etot
+	  // p = rho*(gamma-1)*eint
+	  // cs^2 = gamma * p / rho = gamma*(gamma-1)*eint
+	  enzo_float cs2_1 = std::fmax(0., ggm1*eint_1);
+
+	  // half_factor = 0.5 when eta !=0. Otherwise it's 0.
+	  if ( (cs2_1 > std::fmax(eta*v2, eta*b2*inv_rho)) &&
+	       (eint_1 > half_factor*cur_eint) ){
+	    cur_eint = eint_1;
+	  }
+	  cur_eint = EnzoEquationOfState::apply_floor(cur_eint, eint_floor);
+
+	  eint(iz,iy,ix) = cur_eint;
+	  etot(iz,iy,ix) = cur_eint + kinetic + magnetic;
+	} else {
+
+	  enzo_float etot_floor = eint_floor + kinetic + magnetic;
+	  etot(iz,iy,ix) = EnzoEquationOfState::apply_floor(etot(iz,iy,ix),
+							    etot_floor);
+	}
+      }
+    }
+  }
+}
+
+//----------------------------------------------------------------------
+
+EFlt3DArray EnzoEOSIdeal::retrieve_field_(EnzoFieldArrayFactory &array_factory,
+					  Grouping &group,
+					  std::string group_name, int index,
+					  int reconstructed_axis) const
+{
+  if (reconstructed_axis == -1){
+    return array_factory.from_grouping(group, group_name, index);
+  } else {
+    return array_factory.reconstructed_field(group, group_name, index,
+					     reconstructed_axis);
+  }
+}
+
+//----------------------------------------------------------------------
+
+void EnzoEOSIdeal::copy_passively_advected_fields_
```

Comment:
Is this function still used somewhere?

Reply:
Nope - thanks for pointing that out. I just removed it